src/python/grpcio/grpc/__init__.py

```diff
@@ -45,7 +44,7 @@ class FutureCancelledError(Exception):
     """Indicates that the computation underlying a Future was cancelled."""
 
 
-class Future(six.with_metaclass(abc.ABCMeta)):
```

Comment:
For posterity, since I had to do some research to convince myself that this was a backward compatible change:

- The Cython implementation of [`with_metaclass`](https://github.com/benjaminp/six/blob/3b7efbcca41857da03fb01f004ccc425ab82dfbf/six.py#L856)
- The implementation of [`abc.ABC`](https://github.com/python/cpython/blob/553d3c10172254b190078c50eb9f8e60522c8f41/Lib/abc.py#L184)

I _think_ this should work, but we've heard complaints in the past when making changes to these base classes / metaclasses. Can you please do a little more research and post the results on the PR? What do the old objects look like and what do the new ones look like. Ideally, you should recursively call `dir()` on an instance of the old and an instance of the new and make sure that everything looks the same.

Reply:
Sure, I checked `type` and `dir` for one of the base object `Channel`, here is the result:
#### Context
- [Channel](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_channel.py#L1462) is a class of `grpc.Channel`(internally `grpc._channel.Channel`) which in turn is a [subclass of `abc.ABCMeta`](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/__init__.py#L942).

#### When Using `class Channel(six.with_metaclass(abc.ABCMeta))`:
* Type of `channel`: `<class 'grpc._channel.Channel'>`
* `dir` of `channel`:
```
['__abstractmethods__', '__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_call_state', '_channel', '_close', '_close_on_fork', '_connectivity_state', '_process_python_options', '_single_threaded_unary_stream', '_unsubscribe_all', 'close', 'stream_stream', 'stream_unary', 'subscribe', 'unary_stream', 'unary_unary', 'unsubscribe']
```
* Type of `grpc._channel.Channel`: `<class 'abc.ABCMeta'>`
* `dir` of `grpc._channel.Channel`:
```
['__abstractmethods__', '__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_close', '_close_on_fork', '_process_python_options', '_unsubscribe_all', 'close', 'stream_stream', 'stream_unary', 'subscribe', 'unary_stream', 'unary_unary', 'unsubscribe']
```

#### When Using `class Channel(abc.ABC)`:
* Type of `channel`: `<class 'grpc._channel.Channel'>`
* `dir` of `channel`:
```
['__abstractmethods__', '__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_call_state', '_channel', '_close', '_close_on_fork', '_connectivity_state', '_process_python_options', '_single_threaded_unary_stream', '_unsubscribe_all', 'close', 'stream_stream', 'stream_unary', 'subscribe', 'unary_stream', 'unary_unary', 'unsubscribe']
```
* Type of `grpc._channel.Channel`:`<class 'abc.ABCMeta'>`
* `dir` of `grpc._channel.Channel`:
```
['__abstractmethods__', '__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_close', '_close_on_fork', '_process_python_options', '_unsubscribe_all', 'close', 'stream_stream', 'stream_unary', 'subscribe', 'unary_stream', 'unary_unary', 'unsubscribe']
```
#### Difference
- The only difference between `six.with_metaclass(abc.ABCMeta)` and `abc.ABC` is that both `Channel` and `grpc._channel.Channel` are missing `__slots__` for classes created by Six, that's expected since we can see that the implementation of [abc.ABC](https://github.com/python/cpython/blob/553d3c10172254b190078c50eb9f8e60522c8f41/Lib/abc.py#L184) is adding `__slots__` to `abc.ABCMeta`, I also verified that the `dir()` is the same if we use `metaclass=abc.ABCMeta` instead of `abc.ABC`.


Since use `abc.ABC` is listed as an alternative of `metaclass=abc.ABCMeta` which can help avoid confusing metaclass usage in [official docs](https://docs.python.org/3/library/abc.html#abc.ABC), I think it should be fine to keep it.
Please let me know if you think it's better to use `metaclass=abc.ABCMeta`.

Reply:
:+1: